File: examples/cpp/facade.cpp
#include <iostream>
#include <string>

class Amplifier {
public:
    void on() { std::cout << "  Amplifier on" << std::endl; }
    void setVolume(int level) { std::cout << "  Amplifier volume set to " << level << std::endl; }
    void off() { std::cout << "  Amplifier off" << std::endl; }
};

class DVDPlayer {
public:
    void on() { std::cout << "  DVD Player on" << std::endl; }
    void play(const std::string& movie) { std::cout << "  Playing: " << movie << std::endl; }
    void stop() { std::cout << "  DVD Player stopped" << std::endl; }
    void off() { std::cout << "  DVD Player off" << std::endl; }
};

class Projector {
public:
    void on() { std::cout << "  Projector on" << std::endl; }
    void wideScreenMode() { std::cout << "  Projector in widescreen mode" << std::endl; }
    void off() { std::cout << "  Projector off" << std::endl; }
};

class Lights {
public:
    void dim(int level) { std::cout << "  Lights dimmed to " << level << "%" << std::endl; }
    void on() { std::cout << "  Lights on" << std::endl; }
};

class HomeTheaterFacade {
    Amplifier amp_;
    DVDPlayer dvd_;
    Projector projector_;
    Lights lights_;
public:
    void watchMovie(const std::string& movie) {
        std::cout << "Get ready to watch a movie..." << std::endl;
        lights_.dim(10);
        projector_.on();
        projector_.wideScreenMode();
        amp_.on();
        amp_.setVolume(5);
        dvd_.on();
        dvd_.play(movie);
    }

    void endMovie() {
        std::cout << "Shutting down movie theater..." << std::endl;
        dvd_.stop();
        dvd_.off();
        amp_.off();
        projector_.off();
        lights_.on();
    }
};

int main() {
    std::cout << "=== Facade: Home Theater System ===" << std::endl;

    HomeTheaterFacade theater;
    theater.watchMovie("The Matrix");
    std::cout << std::endl;
    theater.endMovie();

    return 0;
}

File: examples/csharp/Patterns/Facade.cs
namespace DesignPatterns.Facade
{
    public class Amplifier
    {
        public void On() => Console.WriteLine("  Amplifier on");
        public void SetVolume(int level) => Console.WriteLine($"  Amplifier volume set to {level}");
        public void Off() => Console.WriteLine("  Amplifier off");
    }

    public class DvdPlayer
    {
        public void On() => Console.WriteLine("  DVD Player on");
        public void Play(string movie) => Console.WriteLine($"  Playing \"{movie}\"");
        public void Stop() => Console.WriteLine("  DVD Player stopped");
        public void Off() => Console.WriteLine("  DVD Player off");
    }

    public class Projector
    {
        public void On() => Console.WriteLine("  Projector on");
        public void WideScreenMode() => Console.WriteLine("  Projector in widescreen mode");
        public void Off() => Console.WriteLine("  Projector off");
    }

    public class Lights
    {
        public void Dim(int level) => Console.WriteLine($"  Lights dimmed to {level}%");
        public void On() => Console.WriteLine("  Lights on");
    }

    public class HomeTheaterFacade
    {
        private readonly Amplifier _amp = new();
        private readonly DvdPlayer _dvd = new();
        private readonly Projector _projector = new();
        private readonly Lights _lights = new();

        public void WatchMovie(string movie)
        {
            Console.WriteLine($"Get ready to watch \"{movie}\"...");
            _lights.Dim(10);
            _projector.On();
            _projector.WideScreenMode();
            _amp.On();
            _amp.SetVolume(5);
            _dvd.On();
            _dvd.Play(movie);
        }

        public void EndMovie()
        {
            Console.WriteLine("Shutting down...");
            _dvd.Stop();
            _dvd.Off();
            _amp.Off();
            _projector.Off();
            _lights.On();
        }
    }

    public static class Demo
    {
        public static void Run()
        {
            var theater = new HomeTheaterFacade();
            theater.WatchMovie("The Matrix");
            Console.WriteLine();
            theater.EndMovie();
        }
    }
}
